Локализация content/7.blog/212.shiki-v1.md (#48) (#65)
translation-gang/nuxt.com: content/7.blog/21.shiki-v1.md

Чтобы улучшить эту ситуацию, я [начал этот RFC](https://github.com/shikijs/shiki/issues/91), который позже оказался в [этом PR](https://github.com/shikijs/shiki/pull/109) и был отправлен в Shiki v0.9. Хотя он абстрагировал уровень загрузки файлов для использования fetch или файловой системы в зависимости от среды, его все еще довольно сложно использовать, поскольку вам нужно вручную обслуживать грамматики и файлы тем где-то в вашем пакете или CDN, а затем вызывать метод `setCDN`, чтобы сообщить Shiki, куда загружать эти файлы.

Решение не идеальное, но, по крайней мере, оно позволило запустить Shiki в браузере для подсветки динамического контента. С тех пор мы использовали этот подход - до тех пор, пока не началась история этой статьи.

## Начало

> Nuxt прилагает много усилий для продвижения [веба на распределенные серверы](/blog/nuxt-on-the-edge), делая веб более доступным с меньшей задержкой и лучшей производительностью. Как и серверы CDN, службы граничного хостинга, такие как [CloudFlare Workers](https://workers.cloudflare.com/), развернуты по всему миру. Пользователи получают контент с ближайшего сервера без необходимости совершать путешествие на исходный сервер, который может находиться за тысячи миль. Наряду с потрясающими преимуществами, которые они предоставляют, они также идут на некоторые компромиссы. Например, переферийные серверы используют ограниченное runtime-окружение. CloudFlare Workers также не поддерживает доступ к файловой системе и обычно не сохраняет состояние между запросами. Хотя основные накладные расходы Shiki заключаются в предварительной загрузке грамматик и тем, это не будет хорошо работать в переферийной среде.

Reviewer: ```suggestion
Nuxt прилагает много усилий для продвижения [веба на распределенные серверы](/blog/nuxt-on-the-edge), делая веб более доступным с меньшей задержкой и лучшей производительностью. Как и серверы CDN, службы граничного хостинга, такие как [CloudFlare Workers](https://workers.cloudflare.com/), развернуты по всему миру. Пользователи получают контент с ближайшего сервера без необходимости совершать путешествие на исходный сервер, который может находиться за тысячи миль. Наряду с потрясающими преимуществами, которые они предоставляют, они также идут на некоторые компромиссы. Например, периферийные серверы используют ограниченное runtime-окружение. CloudFlare Workers также не поддерживает доступ к файловой системе и обычно не сохраняет состояние между запросами. Хотя основные накладные расходы Shiki заключаются в предварительной загрузке грамматик и тем, это не будет хорошо работать в периферийной среде.
```

> Все началось с чата между [Себастьеном](https://twitter.com/Atinux) и мной. Мы пытались сделать [Nuxt Content](https://github.com/nuxt/content), который использует Shiki для подсветки блоков кода, для работы на переферии.

Reviewer: ```suggestion
Все началось с чата между [Себастьеном](https://twitter.com/Atinux) и мной. Мы пытались сделать [Nuxt Content](https://github.com/nuxt/content), который использует Shiki для подсветки блоков кода, для работы на периферии.
```

![История чата между Себастьеном и Энтони](/assets/blog/shiki-start-chat.png){.rounded-lg.shadow.max-w-[700px].border.dark:border-gray-700}

Я начал эксперименты с локального исправления [`shiki-es`](https://github.com/pi0/shiki-es) (сборка Shiki ESM от [Пуйя Парса](https://github.com/pi0)) для преобразования файлов грамматик и тем в [модуль ECMAScript (ESM)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) так, чтобы его можно было понять и объединить инструментами сборки. Это было сделано для создания пакета кода для CloudFlare Workers, который можно было бы использовать без использования файловой системы или выполнения сетевых запросов.

```ts [До — чтение JSON-ресурсов из файловой системы]
import fs from 'fs/promises'

const cssGrammar = JSON.parse(await fs.readFile('../langs/css.json', 'utf-8'))
```

```ts [После - Использование импорта ESM]
const cssGrammar = await import('../langs/css.mjs').then(m => m.default)
```

> Нам было нужно обернуть JSON-файлы в ESM в виде встроенного литерала, чтобы мы могли использовать [`import()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) для их динамического импорта. Разница в том, что `import()` — это стандартная функция JavaScript, которая работает везде, в то время как `fs.readFile` — это API, специфичный для Node.js, который работает только в Node.js. Наличие `import()` статически также позволило бы сборщикам, таким как [Rollup](https://rollupjs.org/) и [webpack](https://webpack.js.org/), создавать граф взаимосвязей модулей и [выдавать связанный код в виде чанков](https://rollupjs.org/tutorial/#code-splitting).

Reviewer: ```suggestion
Нам нужно было обернуть JSON-файлы в ESM в виде встроенного литерала, чтобы мы могли использовать [`import()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) для их динамического импорта. Разница в том, что `import()` — это стандартная функция JavaScript, которая работает везде, в то время как `fs.readFile` — это API, специфичный для Node.js, который работает только в Node.js. Наличие `import()` статически также позволило бы сборщикам, таким как [Rollup](https://rollupjs.org/) и [webpack](https://webpack.js.org/), создавать граф взаимосвязей модулей и [выдавать связанный код в виде чанков](https://rollupjs.org/tutorial/#code-splitting).
```

Reviewer: так как будто лучше звучит

Конечный результат оказался многообещающим. Нам удалось заставить Shikiji работать в любой среде выполнения, даже с возможностью [импортировать его из CDN и запустить в браузере](https://shiki.style/guide/install#cdn-usage) с помощью одной строки кода.

Мы также воспользовались шансом улучшить API и внутреннюю архитектуру Shiki. Мы перешли от простой конкатенации строк к использованию [`hast`](https://github.com/syntax-tree/hast), создав абстрактное синтаксическое дерево (AST) для генерации выходных данных HTML. Это открывает возможность раскрытия [Transformers API](https://shiki.style/guide/transformers), чтобы позволить пользователям изменять промежуточный HAST и выполнять множество интересных интеграций, которые ранее было бы очень трудно реализовать.

> Поддержка темного/светлого режима [была часто запрашиваемой функцией](https://github.com/shikijs/shiki/issues/33). Из-за статического подхода, который использует Shiki, невозможно было менять тему на лету при рендеринге. Решение в прошлом состояло в том, чтобы дважды генерировать HTML с подсветкой и переключать его видимость на основе предпочтений пользователя — это было неэффективно, так как дублировало payload или использовало [CSS переменные темы](https://github.com/shikijs/shiki/pull/212), что теряло гранулярную подсветку, для которой Shiki отлично подходит. С новой архитектурой Shikiji я сделал шаг назад, переосмыслил проблему  и [придумал идею](https://github.com/shikijs/shiki/issues/33#issuecomment-1676362336) разбить общие токены и объединить несколько тем как встроенные переменные CSS, которые обеспечивают эффективный вывод, при этом согласуясь с философией Shiki. Вы можете узнать больше об этом в [документации Shiki](https://shiki.style/guide/dual-themes).

Reviewer: ```suggestion
Поддержка темного/светлого режима [была часто запрашиваемой функцией](https://github.com/shikijs/shiki/issues/33). Из-за статического подхода, который использует Shiki, невозможно было менять тему на лету при рендеринге. Решение в прошлом состояло в том, чтобы дважды генерировать HTML с подсветкой и переключать его видимость на основе предпочтений пользователя — это было неэффективно, так как дублировало payload или использовало [CSS переменные темы](https://github.com/shikijs/shiki/pull/212), что теряло гранулярную подсветку, для которой Shiki отлично подходит. С новой архитектурой Shikiji я сделал шаг назад, переосмыслил проблему и [придумал идею](https://github.com/shikijs/shiki/issues/33#issuecomment-1676362336) разбить общие токены и объединить несколько тем как встроенные переменные CSS, которые обеспечивают эффективный вывод, при этом согласуясь с философией Shiki. Вы можете узнать больше об этом в [документации Shiki](https://shiki.style/guide/dual-themes).
```

feat!: merge Shikiji back into Shiki for v1.0 [\#557]{.opacity-50}
::

Мы очень рады видеть, что наша работа в Shikiji была влита обратно в Shiki, что не только работает для нас самих, но и приносит пользу всему сообществу. Благодаря этому слиянию, **решается около 95% открытых issues**, которые у нас были в Shiki в течение многих лет:

![Shikiji влит обратно в Shiki](/assets/blog/shiki-merge-pr.png){.rounded-lg.shadow}

Shiki также получил [совершенно новый сайт документации](https://shiki.style/), где вы также можете поиграться с ним прямо в браузере (спасибо агностическому подходу!). Многие фреймворки теперь имеют встроенную интеграцию с Shiki, возможно, вы уже используете его где-то!

## Twoslash

> [Twoslash](https://github.com/twoslashes/twoslash) — это инструмент интеграции для извлечения информации о типе из [TypeScript Language Services](https://github.com/microsoft/TypeScript/wiki/Using-the-Language-Service-API) и генерации в ваш сниппет кода. По сути, он делает статический фрагмент кода похожим на редактор VS Code. Он создан [Orta Therox](https://github.com/orta) для [сайта документации TypeScript](https://github.com/microsoft/TypeScript-Website), вы можете найти [исходный код здесь](https://github.com/microsoft/TypeScript-Website/tree/v2/packages/ts-twoslasher). Orta также создал [интеграцию Twoslash для версий Shiki v0.x](https://github.com/shikijs/twoslash). Тогда у Shiki [не было надлежащей системы плагинов](https://github.com/shikijs/shiki/issues/380), из-за чего `shiki-twoslash` приходилось создавать как оболочку над Shiki, что немного усложняло настройку, поскольку существующие интеграции Shiki не работали напрямую с Twoslash.

Reviewer 2: Orta Therox - это имя и фамилия? не переводим?

Author: @alexeyvokin Наверное нужно перевести, раз везде имена переводим. 
Орта Терокс - так наверное?
Только это мужчина. В таком случае сделан (кем) - Ортой Тероксом?

Мистраль подсказал что вот так:
Транслитерация: Орта Терокс
```
Склонение по падежам:

Именительный падеж: Орта Терокс
Родительный падеж: Орты Терокса
Дательный падеж: Орте Тероксу
Винительный падеж: Орту Терокса
Творительный падеж: Ортой Тероксом
Предложный падеж: Орте Тероксе
```

Reviewer 2: @Hekikai согласен?

Reviewer 3: Да, годно! Оставляем) Поддержу глоссарий в телеге

> Мы также воспользовались шансом пересмотреть интеграции Twoslash, когда переписывали Shikiji, попробовать ["dog-fooding"](https://en.wikipedia.org/wiki/Eating_your_own_dog_food) и проверить расширяемость. С новым внутренним HAST мы можем [интегрировать Twoslash как плагин-трансформер](https://shiki.style/packages/twoslash), что делает его работающим везде, где работает Shiki, а также тспользовать в виде композабла с другими трансформерами.

Reviewer: ```suggestion
Мы также воспользовались шансом пересмотреть интеграции Twoslash, когда переписывали Shikiji, попробовать ["dog-fooding"](https://en.wikipedia.org/wiki/Eating_your_own_dog_food) и проверить расширяемость. С новым внутренним HAST мы можем [интегрировать Twoslash как плагин-трансформер](https://shiki.style/packages/twoslash), что делает его работающим везде, где работает Shiki, а также использовать в виде композабла с другими трансформерами.
```
